src/frontier_exploration/scripts/frontier_filter.py

```diff
@@ -0,0 +1,190 @@
+        self.tfLisn.waitForTransform(
+            global_frame, "/" + self.robot_frame, rospy.Time(0), rospy.Duration(10.0)
+        )
+        rospy.Subscriber(
+            self.goals_topic, PointStamped, callback=self.frontiersCallback
+        )
+        self.frontier_pub = rospy.Publisher("frontier_centroids", Marker, queue_size=10)
+
+        rospy.loginfo("the map and global costmaps are received")
+
+    def frontiersCallback(self, data):
+        x = [array([data.point.x, data.point.y])]
+        if len(self.received_frontiers) > 0:
+            self.received_frontiers = vstack((self.received_frontiers, x))
+        else:
+            self.received_frontiers = x
+
+    def mapCallback(self, data):
+        self.mapData = data
+
+    def init_markers(self):
+        points = Marker()
+        points_clust = Marker()
+        # Set the frame ID and timestamp.  See the TF tutorials for information on these.
+        points.header.frame_id = self.mapData.header.frame_id
+        points.header.stamp = rospy.Time.now()
+        points.ns = "markers2"
+        points.id = 0
+        points.type = Marker.POINTS
+        # Set the marker action for latched self.frontiers.  Options are ADD, DELETE, and new in ROS Indigo: 3 (DELETEALL)
+        points.action = Marker.ADD
+        points.pose.orientation.w = 1.0
+        points.scale.x = 0.2
+        points.scale.y = 0.2
+        points.color.r = 255.0 / 255.0
+        points.color.g = 255.0 / 255.0
+        points.color.b = 0.0 / 255.0
+        points.color.a = 1
+        points.lifetime = rospy.Duration()
+
+        points_clust.header.frame_id = self.mapData.header.frame_id
+        points_clust.header.stamp = rospy.Time.now()
+        points_clust.ns = "markers3"
+        points_clust.id = 4
+
+        points_clust.type = Marker.POINTS
+        # Set the marker action for centroids.  Options are ADD, DELETE, and new in ROS Indigo: 3 (DELETEALL)
+        points_clust.action = Marker.ADD
+        points_clust.pose.orientation.w = 1.0
+        points_clust.scale.x = 0.2
+        points_clust.scale.y = 0.2
+        points_clust.color.r = 0.0 / 255.0
+        points_clust.color.g = 255.0 / 255.0
+        points_clust.color.b = 0.0 / 255.0
+        points_clust.color.a = 1
+        points_clust.lifetime = rospy.Duration()
+
+        return points, points_clust
+
+    def filter(self):
+        # wait if no frontier is received yet
+        rospy.loginfo("Waiting for frontiers")
+        while len(self.received_frontiers) < 1:
+            pass
+
+        points, points_clust = self.init_markers()
+
+        temppoint = PointStamped()
+        temppoint.header.frame_id = self.mapData.header.frame_id
+        temppoint.header.stamp = rospy.Time(0)
+        temppoint.point.z = 0.0
+
+        arraypoints = PointArray()
+        tempPoint = Point()
+        tempPoint.z = 0.0
+
+        p = Point()
+        p.z = 0
+
+        rospy.loginfo("Starting filter")
+        while not rospy.is_shutdown():
```

Comment:
Why is this loop continously running even if received frontiers is empty? Its eating CPU cycles, can you sleep this thread until received frontiers is non empty?

Reply:
I checked its not super straightforward to do it within ROS, one way to do it outside ROS is create a new frontier filter process and then sleep that process on a conditional variable but that will not work in the ROS context,

I ideally instead of creating a separate node for frontier filter which keeps polling if any frontiers are received, a good way to achieve this is through a ROS service which is called by the main frontier node

Reply:
Can let this slide for now!